DXMainClient/Online/Connection.cs

```diff
@@ -220,70 +220,59 @@ private void ConnectToServer()
             connectionManager.OnConnectAttemptFailed();
         }
 
-        private void HandleComm(object client)
+        private void HandleComm()
         {
             int errorTimes = 0;
-
             byte[] message = new byte[1024];
-            int bytesRead;
 
             Register();
 
-            Timer timer = new Timer(new TimerCallback(AutoPing), null, 30000, 120000);
+            Timer timer = new Timer(AutoPing, null, 30000, 120000);
 
             connectionCut = true;
 
             while (true)
             {
-                bytesRead = 0;
-
                 if (connectionManager.GetDisconnectStatus())
                 {
                     connectionManager.OnDisconnected();
                     connectionCut = false; // This disconnect is intentional
                     break;
                 }
 
+                if (!serverStream.DataAvailable)
+                {
+                    Thread.Sleep(10);
+                    continue;
+                }
+
+                int bytesRead;
+
                 try
                 {
                     bytesRead = serverStream.Read(message, 0, 1024);
                 }
-                catch (Exception ex)
+                catch (IOException ex) when (IsKnownSocketException(ex))
                 {
                     errorTimes++;
-
-                    if (errorTimes > 30) // TODO Figure out if this hacky check is actually necessary
-                    {
-                        Logger.Log("Disconnected from CnCNet due to a socket error. Message: " + ex.Message);
-                        failedServerIPs.Add(currentConnectedServerIP);
-                        connectionManager.OnConnectionLost(ex.Message);
-                        break;
-                    }
-                    else if (connectionManager.GetDisconnectStatus())
-                    {
-                        connectionManager.OnDisconnected();
-                        connectionCut = false; // This disconnect is intentional
-                        break;
-                    }
-
                     continue;
                 }
-
-                if (bytesRead == 0)
+                catch (Exception ex)
                 {
+                    Logger.Log("Disconnected from CnCNet due to a socket error. Message: " + ex.Message);
                     errorTimes++;
-
-                    if (errorTimes > 30) // TODO Figure out if this hacky check is actually necessary
-                    {
-                        failedServerIPs.Add(currentConnectedServerIP);
-                        Logger.Log("Disconnected from CnCNet.");
-                        connectionManager.OnConnectionLost("Server disconnected.".L10N("UI:Main:ServerDisconnected"));
-                        break;
-                    }
-
                     continue;
                 }
 
+                if (errorTimes > 30)
```

Comment:
Would be good to make this into a const since you touch that part of the code anyways.

Reply:
@Metadorius I reused the existing MAX_RECONNECT_COUNT = 8 since 30 is very excessive. Exceptions will rarely be thrown with the new code.